# Document Chunking Strategies for RAG

## Deliverables
- Chunking strategy comparison
- Example implementations
- Retrieval quality observations
- Best practices

## 1. Imports

In [1]:
from textwrap import wrap
import re

sample_text = '''
Retrieval-Augmented Generation (RAG) improves LLM responses by retrieving relevant documents from a knowledge base.
Document chunking is the process of splitting large documents into smaller pieces before generating embeddings.
Different chunking strategies affect retrieval quality, context preservation, and storage cost.
Overlapping chunks preserve context, while semantic chunking tries to keep related ideas together.
''' * 5

print(sample_text[:300])


Retrieval-Augmented Generation (RAG) improves LLM responses by retrieving relevant documents from a knowledge base.
Document chunking is the process of splitting large documents into smaller pieces before generating embeddings.
Different chunking strategies affect retrieval quality, context preserv


## 2. Fixed-Size Chunking

In [2]:
def fixed_size_chunk(text, size=200):
    return wrap(text, size)

chunks = fixed_size_chunk(sample_text, 200)
print('Total Chunks:', len(chunks))
print(chunks[0])

Total Chunks: 11
 Retrieval-Augmented Generation (RAG) improves LLM responses by retrieving relevant documents from a knowledge base. Document chunking is the process of splitting large documents into smaller pieces


## 3. Sliding Window Chunking

In [3]:
def sliding_window_chunk(text, chunk_size=200, overlap=50):
    chunks=[]
    start=0
    while start < len(text):
        chunks.append(text[start:start+chunk_size])
        start += chunk_size-overlap
    return chunks

chunks=sliding_window_chunk(sample_text)
print(len(chunks))
print(chunks[0])
print(chunks[1])

15

Retrieval-Augmented Generation (RAG) improves LLM responses by retrieving relevant documents from a knowledge base.
Document chunking is the process of splitting large documents into smaller pieces b
of splitting large documents into smaller pieces before generating embeddings.
Different chunking strategies affect retrieval quality, context preservation, and storage cost.
Overlapping chunks preser


## 4. Sentence-Based Chunking

In [4]:
def sentence_chunk(text, n=2):
    s=re.split(r'(?<=[.!?])\s+', text.strip())
    return [' '.join(s[i:i+n]) for i in range(0,len(s),n)]

chunks=sentence_chunk(sample_text)
print(len(chunks))
print(chunks[0])

10
Retrieval-Augmented Generation (RAG) improves LLM responses by retrieving relevant documents from a knowledge base. Document chunking is the process of splitting large documents into smaller pieces before generating embeddings.


## 5. Paragraph-Based Chunking

In [5]:
paragraphs=[p.strip() for p in sample_text.split('\n\n') if p.strip()]
print('Paragraphs:', len(paragraphs))
print(paragraphs[0][:200])

Paragraphs: 5
Retrieval-Augmented Generation (RAG) improves LLM responses by retrieving relevant documents from a knowledge base.
Document chunking is the process of splitting large documents into smaller pieces be


## 6. Simple Semantic Chunking

In [6]:
sentences=re.split(r'(?<=[.!?])\s+', sample_text)
semantic=[]
current=[]
for s in sentences:
    if 'retrieval' in s.lower() or 'rag' in s.lower():
        current.append(s)
    else:
        if current:
            semantic.append(' '.join(current))
            current=[]
        semantic.append(s)
if current:
    semantic.append(' '.join(current))
print('Semantic chunks:', len(semantic))
print(semantic[0])

Semantic chunks: 21

Retrieval-Augmented Generation (RAG) improves LLM responses by retrieving relevant documents from a knowledge base.


## 7. Strategy Comparison

| Strategy | Pros | Cons | Best Use |
|---|---|---|---|
| Fixed Size | Fast | Context loss | Simple RAG |
| Sliding Window | Better continuity | More storage | QA |
| Sentence | Natural | Variable size | Articles |
| Paragraph | Logical | Uneven sizes | Documentation |
| Semantic | Best retrieval | Computationally expensive | Enterprise RAG |


## 8. Retrieval Quality Observations

- Small chunks: high precision, low context.
- Large chunks: rich context, lower precision.
- Overlap improves continuity but increases storage.
- Semantic chunking gives the best retrieval quality in many cases.


## 9. Best Practices

1. Use 400–600 token chunks as a starting point.
2. Use 10–15% overlap.
3. Preserve complete ideas.
4. Store metadata.
5. Evaluate chunking with retrieval metrics.
